### PR TITLE
Resource Manager impl

### PR DESCRIPTION
**Pull request by JerryImMouse**
?

### CODE REVIEW

File: Hypercube.Shared/Resources/ResourcePath.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
+﻿using System.Diagnostics.CodeAnalysis;
+
+namespace Hypercube.Shared.Resources;
+
+public struct ResourcePath
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/ResourcePath.cs:5]: May be readonly

File: Hypercube.Shared/Resources/Manager/IResourceManager.cs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+﻿namespace Hypercube.Shared.Resources.Manager;
+
+public class IResourceManager
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/Manager/IResourceManager.cs:3]: It's interface

File: Hypercube.Shared/Resources/ResourcePath.cs [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+                ? filename
+                : filename[..ind];
+        }
+    }
+
+    public ResourcePath ParentDirectory
+    {
+        get
+        {
+            if (IsSelf)
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/ResourcePath.cs:71]:
```cs
if (IsSelf)
    return Self;
```

File: Hypercube.Shared/Resources/ResourcePath.cs [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    public static ResourcePath operator +(ResourcePath l, ResourcePath r)
+    {
+        if (r.IsSelf)
+            return l;
+        if (r.Rooted)
+            return r;
+        if (l.Path == "")
+            return new ResourcePath("/" + r.Path);
+
+        if (l.Path.EndsWith("/"))
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/ResourcePath.cs:103]:
```cs
if (l.Path.EndsWith("/"))
    return new ResourcePath(l.Path + r.Path);
```

File: Hypercube.Shared/Resources/ResourcePath.cs [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            return idx switch
+            {
+                -1 => Self,
+                0 => new ResourcePath(Path[..1]),
+                _ => new ResourcePath(Path[..idx])
+            };
+        }
+    }
+
+    public static implicit operator ResourcePath(string path)
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/ResourcePath.cs:89]: Make implicit for string from ResourcePath

File: Hypercube.Shared/Resources/ResourcePath.cs [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        return new ResourcePath(path);
+    }
+    
+    public static ResourcePath operator +(ResourcePath l, ResourcePath r)
+    {
+        if (r.IsSelf)
+            return l;
+        if (r.Rooted)
+            return r;
+        if (l.Path == "")
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/ResourcePath.cs:100]: replcae `""` to `string.Empty`

File: Hypercube.Shared/Resources/ResourcePath.cs [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+            return new ResourcePath("/" + r.Path);
+
+        if (l.Path.EndsWith("/"))
+        {
+            return new ResourcePath(l.Path + r.Path);
+        }
+
+        return new ResourcePath(l.Path + "/" + r.Path);
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/ResourcePath.cs:108]: Maybe `$"{l.Path }/{r.Path}"`?

**Review by Tornado-Technology — CHANGES_REQUESTED** [review]

File: Hypercube.Shared/Resources/ResourcePath.cs (hunk 5 above)
Comment (JerryImMouse) [reply Hypercube.Shared/Resources/ResourcePath.cs:97]: done

File: Hypercube.Shared/Resources/ResourcePath.cs (hunk 7 above)
Comment (JerryImMouse) [reply Hypercube.Shared/Resources/ResourcePath.cs]: done

File: Hypercube.Shared/Resources/ResourcePath.cs (hunk 6 above)
Comment (JerryImMouse) [reply Hypercube.Shared/Resources/ResourcePath.cs]: done

File: Hypercube.Shared/Resources/ResourcePath.cs (hunk 4 above)
Comment (JerryImMouse) [reply Hypercube.Shared/Resources/ResourcePath.cs:115]: done

File: Hypercube.Shared/Resources/ResourcePath.cs (hunk 3 above)
Comment (JerryImMouse) [reply Hypercube.Shared/Resources/ResourcePath.cs:79]: done

File: Hypercube.Shared/Resources/Manager/IResourceManager.cs (hunk 2 above)
Comment (JerryImMouse) [reply Hypercube.Shared/Resources/Manager/IResourceManager.cs]: erghhh...

File: Hypercube.Shared/Resources/ResourcePath.cs (hunk 1 above)
Comment (JerryImMouse) [reply Hypercube.Shared/Resources/ResourcePath.cs]: done

File: Hypercube.Client/Graphics/Texturing/TextureManager.cs [hunk 8]
```diff
@@ -1,21 +1,25 @@
...
         StbImage.stbi_set_flip_vertically_on_load(1);
     }
     
-    public ITexture Create(string path)
+    public ITexture Create(ResourcePath path)
     {
-        using var stream = File.OpenRead(path);
-        return Create(ImageResult.FromStream(stream, ColorComponents.RedGreenBlueAlpha));
+        return Create(ImageResult.FromStream(_resMan.ReadFileContent(path) ?? throw new FileNotFoundException(), ColorComponents.RedGreenBlueAlpha));
     }
```
Comment (Tornado-Technology) [inline Hypercube.Client/Graphics/Texturing/TextureManager.cs:21]: Do flip joke died. noooooo

File: Hypercube.Client/Graphics/Windows/Manager/GlfwWindowManager.Window.cs [hunk 9]
```diff
@@ -186,9 +188,9 @@ private bool TryGetWindow(Window* window, [NotNullWhen(true)] out GlfwWindowRegi
          return null;
      }
 
-     public IEnumerable<ITexture> LoadWindowIcon(ITextureManager textureMan, string resPath)
+     public IEnumerable<ITexture> LoadWindowIcon(ITextureManager textureMan, IResourceManager resourceMan, ResourcePath path)
```
Comment (Tornado-Technology) [inline Hypercube.Client/Graphics/Windows/Manager/GlfwWindowManager.Window.cs:191]: image

File: Hypercube.Shared/Resources/DirRoot/DirContentRoot.cs [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+            return true;
+        }
+        catch (Exception ex)
+        {
+            _logger.Error(ex.Message);
+            stream = null;
+        }
+        return false;
+    }
+    private bool FileExists(ResourcePath relPath)
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/DirRoot/DirContentRoot.cs:38]: Where my indents

File: Hypercube.Client/Graphics/Windows/Manager/IWindowManager.cs [hunk 11]
```diff
@@ -30,7 +32,7 @@ public interface IWindowManager : IDisposable
     void WindowSetVisible(WindowRegistration registration, bool visible);
     void WindowSetSize(WindowRegistration registration, Vector2Int size);
     void WindowSwapBuffers(WindowRegistration window);
-    IEnumerable<ITexture> LoadWindowIcon(ITextureManager textureMan, string resPath);
+    IEnumerable<ITexture> LoadWindowIcon(ITextureManager textureMan, IResourceManager resourceManager, ResourcePath resPath);
```
Comment (Tornado-Technology) [inline Hypercube.Client/Graphics/Windows/Manager/IWindowManager.cs:35]: TexutreMan so funny

File: Hypercube.Shared/Resources/Manager/ResourceManager.cs [hunk 12]
```diff
@@ -0,0 +1,140 @@
+﻿using System.Diagnostics.CodeAnalysis;
+using Hypercube.Shared.Logging;
+using Hypercube.Shared.Resources.DirRoot;
+using Hypercube.Shared.Utilities.Helpers;
+
+namespace Hypercube.Shared.Resources.Manager;
+
+public class ResourceManager : IResourceManager
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/Manager/ResourceManager.cs:8]: sealed

File: Hypercube.Shared/Resources/Manager/ResourceManager.cs [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+{
+    private Logger _logger = LoggingManager.GetLogger("resources");
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/Manager/ResourceManager.cs:10]: readonly

File: Hypercube.Shared/Resources/Manager/ResourceManager.cs [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+    private (ResourcePath prefix, IContentRoot root)[] _roots = [];
+    private readonly object _rootLock = new();
+    public void AddRoot(ResourcePath prefix, IContentRoot root)
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/Manager/ResourceManager.cs:13]: indent

File: Hypercube.Shared/Resources/Manager/ResourceManager.cs [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+    public void MountContentFolder(string file, ResourcePath? prefix = null)
+    {
+        prefix = ValidatePrefix(prefix);
+        
+        if (!Path.IsPathRooted(file))
+            file = PathHelpers.GetExecRelativeFile(file);
+
+        var dirInfo = new DirectoryInfo(file);
+        if (!dirInfo.Exists)
+        {
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/Manager/ResourceManager.cs:33]: brakeless

File: Hypercube.Shared/Resources/Manager/ResourceManager.cs [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+            _logger.Error(ex.Message);
+        }
+
+        fileStream = null;
+        return false;
+    }
+
+    public IEnumerable<ResourcePath> FindContentFiles(ResourcePath? path)
+    {
+        if (path == null)
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/Manager/ResourceManager.cs:77]: `path is null`

File: Hypercube.Shared/Resources/Manager/ResourceManager.cs [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+        {
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/Manager/ResourceManager.cs:78]: brakeless

File: Hypercube.Shared/Resources/Manager/ResourceManager.cs [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+            throw new ArgumentNullException(nameof(path));
+        }
+
+        if (!path.Value.Rooted)
+        {
+            throw new ArgumentException("Path is not rooted", nameof(path));
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/Manager/ResourceManager.cs:84]: brakeless

File: Hypercube.Shared/Resources/Manager/ResourceManager.cs [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+        }
+        
+        var returned = new HashSet<ResourcePath>();
+        
+        foreach (var (prefix, root) in _roots)
+        {
+            if (!path.Value.TryRelativeTo(prefix, out var relative))
+            {
+                continue;
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/Manager/ResourceManager.cs:93]: brakeless

File: Hypercube.Shared/Resources/Manager/ResourceManager.cs [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+            }
+
+            foreach (var filename in root.FindFiles(relative.Value))
+            {
+                var newPath = prefix + filename;
+                if (returned.Add(newPath))
+                {
+                    yield return newPath;
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/Manager/ResourceManager.cs:101]: brakeless

File: Hypercube.Shared/Resources/Manager/ResourceManager.cs [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+                }
+            }
+        }
+    }
+
+    public string ReadFileContentAllText(ResourcePath path)
+    {
+        var stream = ReadFileContent(path);
+        if (stream == null)
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/Manager/ResourceManager.cs:110]: `stream is null`

File: Hypercube.Shared/Resources/Manager/ResourceManager.cs [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+        if (!TryReadFileContent(path, out var stream))
+            return null;
+
+        return stream;
+    }
+
+    private ResourcePath ValidatePrefix(ResourcePath? prefix)
+    {
+        if (prefix is null)
+            prefix = "/";
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/Manager/ResourceManager.cs:127]: return here?

File: Hypercube.Shared/Resources/ResourcePath.cs [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+        if (OperatingSystem.IsWindows())
+        {
+            Path = path.Replace('\\', '/');
+            return;
+        }
+
+        Path = path;
+    }
+
+    static ResourcePath()
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Resources/ResourcePath.cs:28]: Move under that `public ResourcePath`

File: Hypercube.Shared/Utilities/Helpers/PathHelpers.cs [hunk 24]
```diff
@@ -0,0 +1,25 @@
...
+    {
+        return Path.GetFullPath(Path.Combine(GetExecDirectory(), file));
+    }
+    
+    public static IEnumerable<string> GetFiles(string path)
+    {
+        return Directory.EnumerateFiles(path, "*", SearchOption.AllDirectories);
+    }
+    
+    public static bool IsFileSystemCaseSensitive() =>
```
Comment (Tornado-Technology) [inline Hypercube.Shared/Utilities/Helpers/PathHelpers.cs:22]: Use with body

**Review by Tornado-Technology — CHANGES_REQUESTED** [review]
Much better

File: Hypercube.Client/Graphics/Texturing/TextureManager.cs (hunk 8 above)
Comment (JerryImMouse) [reply Hypercube.Client/Graphics/Texturing/TextureManager.cs:21]: its in ITextureManager, not in TextureManager, i can use // <inheritdoc>, i think

File: Hypercube.Client/Graphics/Windows/Manager/GlfwWindowManager.Window.cs (hunk 9 above)
Comment (JerryImMouse) [reply Hypercube.Client/Graphics/Windows/Manager/GlfwWindowManager.Window.cs]: 😢

**Review by Tornado-Technology — APPROVED** [review]
Yapiiii

**Pull request merged**